# lorals_paper_code
Data tables and scripts for <i>Transcriptome variation in human tissues revealed by long-read sequencing</i>.

## Overview of repository
<ul>
  <li><a href="https://github.com/LappalainenLab/lorals_paper_code/tree/master/data">data/</a>: Publicly available data used in this pipeline </li> 
  <li><a href="https://github.com/LappalainenLab/lorals_paper_code/tree/master/analysis">analysis/</a>: Data tables organized by analysis generated throughout the pipeline (some small tables are included) </li> 
</ul>

## Analyses included
<ul>
  <li><b>Qualitative control</b>:</li>
  <ul>
    <li>Qualitative control by NanoPlot</li>
    <li>Three prime bias using the mitochondrial transcripts</li>
    <li>Overview of the dataset aligned to GENCODEv26</li>
  </ul>
  <li><b>Call new transcript with flair</b>:</li>
  <ul>
    <li>Flair and TransDecoder pipeline to call novel transcripts</li>
    <li>Comparison of FLAIR novel transcripts to other datasets</li>
    <li>Overview of the dataset aligned to FLAIR</li>
  </ul>
  <li><b>Differential transcripts across tissues</b>:</li>
  <ul>
    <li>Differential transcript expression and usage across tissues</li>
  </ul>
  <li><b>Allelic analysis</b>:</li>
  <ul>
    <li>Analysis of allelic counts from genes and transcripts</li>
    <li>Comparison to GTEx eQTLs and sQTLs</li>
    <li>Stratification by alternative transcript structure event</li>
    <li>Analysis of enloc</li>
  </ul>
  <li><b>PTBP1 knockdown</b>:</li>
  <ul>
    <li>Allele analysis in fibroblast cell lines with PTBP1 KD</li>
    <li>Comparison to ENCODE eCLIP sites</li>
  </ul>
  <li><b>Variant interpretation and rare variant analysis</b>:</li>
  <ul>
    <li>Variant annotation using new flair transcripts</li>
    <li>Rare variant interpretation using new flair transcripts</li>
  </ul>
</ul>
